crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2615,36 +2615,112 @@ impl<'db> Type<'db> {
+                            // the example above.
+                            KnownClass::Bool => arguments
+                                .first_argument()
+                                .map(|arg| arg.bool(db).into_type(db))
+                                .unwrap_or(Type::BooleanLiteral(false)),
+
+                            // TODO: Don't ignore the second and third arguments to `str`
+                            //   https://github.com/astral-sh/ruff/pull/16161#discussion_r1958425568
+                            KnownClass::Str => arguments
+                                .first_argument()
+                                .map(|arg| arg.str(db))
+                                .unwrap_or_else(|| Type::string_literal(db, "")),
+
+                            KnownClass::Type => arguments
+                                .exactly_one_argument()
+                                .map(|arg| arg.to_meta_type(db))
+                                .unwrap_or_else(|| KnownClass::Type.to_instance(db)),
+
+                            // TODO: remove a dedicated branch when `__new__` handling is added.
+                            // at that point the regular logic should be able to check signatures
+                            // of other KnownClass constructors against typeshed.
+                            _ => Type::Instance(InstanceType { class }),
+                        },
+                    )))
+                } else {
+                    // User defined class
+                    let instance_ty = Type::Instance(InstanceType { class });
+
+                    // Try calling __init__ to check arguments before returning the instance type
+                    match instance_ty.try_call_dunder(db, "__init__", arguments) {
+                        Err(CallDunderError::Call(mut error)) => {
+                            // If the call to `__init__` fails we want to return inner error directly
+                            // but ensure that the fallback return type is the instance type.
+
+                            match error {
+                                CallError::NotCallable { .. } => {
+                                    // Instead of trying to emit a diagnostic at call site, these should be
+                                    // caught at the definition site of the class. Here we return Ok with the
+                                    // instance type to ensure that the return type is correct.
+                                    Ok(CallOutcome::Single(CallBinding::from_return_type(
+                                        instance_ty,
+                                    )))
+                                }
+                                CallError::PossiblyUnboundDunderCall { .. } => {
+                                    tracing::warn!("Inferred `__init__` as possibly unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
```

Comment:
This case would not happen due to possibly unbound `__init__` -- that would trigger `CallDunderError::PossiblyUnbound` (which is handled below.) This case here, one layer deeper, would be triggered if, for example, we have an object set as `__init__` which has a possibly-unbound `__call__` method. This can happen without a broken typeshed, so I think in this case we should `set_return_type` and return `Err(error)` just like in the next case, so that we emit a diagnostic. (And add a test for this scenario.)